azurelinuxagent/common/utils/networkutil.py

```diff
@@ -226,43 +210,62 @@ def __execute_cmd(cmd):
+               and AddFirewallRules.__execute_check_command(check_cmd_drop_rule)
 
     @staticmethod
-    def __execute_firewalld_commands(command, dst_ip, uid):
+    def __execute_firewall_commands(dst_ip, uid, command=APPEND_COMMAND, firewalld_command="", wait=""):
+        # The order in which the below rules are added matters for the ip table rules to work as expected
+
         AddFirewallRules.__raise_if_empty(dst_ip, "Destination IP")
         AddFirewallRules.__raise_if_empty(uid, "User ID")
 
         # Firewalld.service fails if we set `-w` in the iptables command, so not adding it at all for firewalld commands
 
-        accept_tcp_rule = AddFirewallRules.get_accept_tcp_rule(AddFirewallRules.INSERT_COMMAND, dst_ip, firewalld_command=command)
+        accept_tcp_rule = AddFirewallRules.get_accept_tcp_rule(command, dst_ip,
+                                                               firewalld_command=firewalld_command, wait=wait)
         AddFirewallRules.__execute_cmd(accept_tcp_rule)
 
-        accept_cmd = AddFirewallRules.get_firewalld_accept_command(command, dst_ip, uid)
+        accept_cmd = AddFirewallRules.get_accept_command(command, dst_ip, uid,
+                                                         firewalld_command=firewalld_command, wait=wait)
         AddFirewallRules.__execute_cmd(accept_cmd)
 
-        drop_cmd = AddFirewallRules.get_firewalld_drop_command(command, dst_ip)
+        drop_cmd = AddFirewallRules.get_drop_command(command, dst_ip, firewalld_command=firewalld_command, wait=wait)
         AddFirewallRules.__execute_cmd(drop_cmd)
 
+    @staticmethod
+    def add_iptables_rules(wait, dst_ip, uid):
+        AddFirewallRules.__execute_firewall_commands(dst_ip, uid, command=AddFirewallRules.APPEND_COMMAND, wait=wait)
+
     @staticmethod
     def add_firewalld_rules(dst_ip, uid):
         # Firewalld.service fails if we set `-w` in the iptables command, so not adding it at all for firewalld commands
         # Firewalld.service with the "--permanent --passthrough" parameter ensures that a firewall rule is set only once even if command is executed multiple times
 
-        AddFirewallRules.__execute_firewalld_commands(FirewallCmdDirectCommands.PassThrough, dst_ip, uid)
+        AddFirewallRules.__execute_firewall_commands(dst_ip, uid, firewalld_command=FirewallCmdDirectCommands.PassThrough)
 
     @staticmethod
     def check_firewalld_rule_applied(dst_ip, uid):
-        AddFirewallRules.__execute_firewalld_commands(FirewallCmdDirectCommands.QueryPassThrough, dst_ip, uid)
+        AddFirewallRules.__execute_firewall_commands(dst_ip, uid, firewalld_command=FirewallCmdDirectCommands.QueryPassThrough)
+
+    @staticmethod
+    def remove_firewalld_rules(dst_ip, uid):
+        AddFirewallRules.__execute_firewall_commands(dst_ip, uid, firewalld_command=FirewallCmdDirectCommands.RemovePassThrough)
```

Comment:
Not sure if you already did this check, but did you verify how `firewalld` handles it if we try to `remove-passthrough` a command that does not exist? Depending on if it exits with a non-0 exit code, we might need to handle it properly in the `persist_firewall_rules.py @ ln: 53`

Reply:
It's always return success if rule exist or not. If we think it may return other command errors or error codes, then it's true for `passthrough `or `query-passthrough`. We may need address for all.

Reply:
>  If we think it may return other command errors or error codes

It's not a guessing game, we should verify from the code and the VMs on how the service actually reacts to it :)

> then it's true for passthrough or query-passthrough

Not sure I follow you there. Both those commands have a different use case altogether. `passthrough` just adds iptable rules using firewalld. `query-passthrough` checks if a rule is there or not, if not it exits with a non-0 code and raises and we have exception handling around that code to handle those scenarios. 

Similarly, I want to verify 2 things if we try removing an iptable rule from firewalld service using `remove-passthrough` - 
- Do the rules get removed (adding a test here would help if you can)
- If we try to remove a rule that does not exist, does it fail with a non-0 exit code or not. Because if it does indeed fail with a non-0 exit code, the implementation that you've added will not handle that case

Reply:
Might also be worthwhile to add a test case for this case too (in case `remove_firewalld_rules` fails and raises an exception because you haven't added any handling there)

Reply:
As I said before, as per my testing `remove-passthrough` returns success regardless rule exist or not. I also removed the rule and restart the agent. Agent didn't raise any sort of errors. 

Same test I added in Dcr too.

Reply:
Having said that, at the end of the day those are shell commands and may return exit codes between1-255. So, I handled it now.

Reply:
Thanks